gtime/model_selection/cross_validation.py

```diff
@@ -0,0 +1,99 @@
+import pandas as pd
+
+
+def time_series_split(time_series: pd.DataFrame, n_splits=4, split_on='index'):
+   """
+   Split the input DataFrame into n_splits. If the data is not a timeries then the split
+   is based on the number of samples.
+   If the data is a timeseries and split_on 'time' then divide the time series based on
+   time.
+
+   Note: The split is based on the index, if split_on is timeseries then the data will be split 
+   based on time
+
+   Parameters
+   ----------
+   time_series : pandas DataFrame, shape (n_samples, n_features), required
+
+   n_splits : int, default = 4, required
+   The number of splits/folds on the dataset
+
+   split_on : 'index', default = 'index'. Optional - 'time' 
+   If the parameter is 'time' then DataFrame index must be DateTime. The dataset will be split based on time
+
+   Yields
+   -------
+   fold.index : RangeIndex indexes of folds 
+   time_fold.index : DateTimeIndex of folds if split_on 'time' 
+
+   Examples
+   --------
+   Example 1)
+   >>> date_rng = pd.date_range(start='1/1/2018', end='1/08/2018', freq='D')
+   >>> time_series = pd.DataFrame(date_rng, columns=['date'])
+   >>> time_series.set_index('date', inplace=True)
+   >>> time_series['data'] = np.random.randint(0,100,size=(len(date_rng)))
+   >>> for fold in (time_series_split(time_series, n_splits=4, split_on='time')):
+   ...      print(fold)
+   DatetimeIndex(['2018-01-01', '2018-01-02'], dtype='datetime64[ns]', name='date', freq=None)
+   DatetimeIndex(['2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04'], 
+                  dtype='datetime64[ns]', name='date', freq=None)
+   DatetimeIndex(['2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04',
+                  '2018-01-05', '2018-01-06'],
+                   dtype='datetime64[ns]', name='date', freq=None)
+   DatetimeIndex(['2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04',
+                  '2018-01-05', '2018-01-06', '2018-01-07', '2018-01-08'],
+                   dtype='datetime64[ns]', name='date', freq=None)
+
+   Example 2)
+   >>> df = pd.DataFrame(np.random.randint(0,100,size=(16, 4)), columns=list('ABCD'))
+   >>> for fold in (time_series_split(df, n_splits=4, split_on='index')):
+   ...      print(fold)
+   RangeIndex(start=0, stop=4, step=1)
+   RangeIndex(start=0, stop=8, step=1)
+   RangeIndex(start=0, stop=12, step=1)
+   RangeIndex(start=0, stop=16, step=1)
+
+   """
+   
+   n_samples = len(time_series)
+   if n_samples < n_splits:
+      raise ValueError(
+         "The number of splits is greater than number of samples"
+      )
+   
```

Comment:
I would create two subfunctions `_time_series_split_on_index` and `_time_series_split_on_time` that are called in the if/elif/else

Reply:
The suggested approach surely looks clean and makes the code organised but don't you think it will add more time complexity? as making them as subfunctions and calling them from if/elif will need for loop to fetch the yields from those subfunctions.

Reply:
Just call the function inside the `if`, without the for loop, it should not cause execution time problems.
And can you put the two functions outside of the main one?

Reply:
Used `yield from` in if/elif to address this issue. The issue is now resolved.